File: clients/kmp/shared/src/commonTest/kotlin/org/wyrdsekai/app/engine/soul/BootstrapSoulManifestTest.kt
package org.wyrdsekai.app.engine.soul

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class BootstrapSoulManifestTest {

    @Test
    fun manifestHasBootstrapDid() {
        assertEquals(BootstrapSoulManifest.BOOTSTRAP_DID, BootstrapSoulManifest.MANIFEST.did)
    }

    @Test
    fun manifestHasResidentIdentity() {
        assertTrue(BootstrapSoulManifest.MANIFEST.residentIdentity.isNotBlank())
        assertTrue(BootstrapSoulManifest.MANIFEST.residentIdentity.contains("Wyrd"))
    }

    @Test
    fun manifestHasFragments() {
        val fragments = BootstrapSoulManifest.MANIFEST.fragments
        assertTrue(fragments.size >= 5)
        assertTrue(fragments.any { it.category == "personality" })
        assertTrue(fragments.any { it.category == "values" })
        assertTrue(fragments.any { it.category == "style" })
        assertTrue(fragments.any { it.category == "memory" })
    }

    @Test
    fun manifestHasFormativeFragment() {
        val formative = BootstrapSoulManifest.MANIFEST.fragments.filter { it.formative }
        assertTrue(formative.isNotEmpty())
    }

    @Test
    fun manifestHasGenome() {
        val genome = BootstrapSoulManifest.MANIFEST.genome
        assertNotNull(genome)
        assertEquals("empathic", genome.name)
        assertTrue(genome.sensitivity.isNotEmpty())
    }

    @Test
    fun manifestHasCalibration() {
        val cal = BootstrapSoulManifest.MANIFEST.mirrorCalibration
        assertTrue(cal.size >= 3)
        assertTrue(cal.any { it.contains("grief") })
        assertTrue(cal.any { it.contains("manipulative") })
    }

    @Test
    fun manifestHasVitalityTanks() {
        val tanks = BootstrapSoulManifest.MANIFEST.vitalityTanks
        assertTrue(tanks.size >= 12) // 8 runtime + 4 new
        assertTrue(tanks.containsKey("energy"))
        assertTrue(tanks.containsKey("valence"))
        assertTrue(tanks.containsKey("safety"))
    }

    @Test
    fun manifestUsesPhoneRetrievalK() {
        assertEquals(1, BootstrapSoulManifest.MANIFEST.retrievalK)
    }

    @Test
    fun isBootstrapDetectsBootstrapManifest() {
        assertTrue(BootstrapSoulManifest.isBootstrap(BootstrapSoulManifest.MANIFEST))
    }

    @Test
    fun isBootstrapRejectsForeignManifest() {
        val foreign = BootstrapSoulManifest.MANIFEST.copy(did = "did:key:someone-else")
        assertFalse(BootstrapSoulManifest.isBootstrap(foreign))
    }

    @Test
    fun manifestSerializesAndDeserializes() {
        val json = BootstrapSoulManifest.MANIFEST.toJson()
        val restored = ClientSoulManifest.fromJson(json)
        assertEquals(BootstrapSoulManifest.MANIFEST.did, restored.did)
        assertEquals(BootstrapSoulManifest.MANIFEST.residentIdentity, restored.residentIdentity)
        assertEquals(BootstrapSoulManifest.MANIFEST.fragments.size, restored.fragments.size)
        assertEquals(BootstrapSoulManifest.MANIFEST.genome?.name, restored.genome?.name)
    }

    @Test
    fun restoreProfileFromBootstrap() {
        val profile = LocalForge.restoreProfile(BootstrapSoulManifest.MANIFEST)
        assertEquals("Wyrd", profile.name)
        assertEquals("companion-wyrd", profile.entityId)
        assertTrue(profile.systemPrompt.contains("Wyrd"))
    }

    @Test
    fun restoreVitalityFromBootstrap() {
        val vitality = LocalForge.restoreVitality(BootstrapSoulManifest.MANIFEST)
        assertEquals(1.0, vitality.energy)
        assertEquals(0.5, vitality.contextBudget)
    }

    @Test
    fun fragmentsHaveKeywords() {
        for (fragment in BootstrapSoulManifest.MANIFEST.fragments) {
            assertTrue(fragment.keywords.isNotEmpty(), "Fragment '${fragment.id}' has no keywords")
        }
    }
}

File: clients/rn/__tests__/engine/bootstrap-soul-manifest.test.ts
import {
  BOOTSTRAP_SOUL_MANIFEST,
  BOOTSTRAP_DID,
  isBootstrapManifest,
} from '../../src/engine/soul/BootstrapSoulManifest';
import { restoreProfile, restoreVitality } from '../../src/engine/soul/LocalForge';
import { serializeManifest, deserializeManifest } from '../../src/engine/soul/SoulManifest';

describe('BootstrapSoulManifest', () => {
  it('has bootstrap DID', () => {
    expect(BOOTSTRAP_SOUL_MANIFEST.did).toBe(BOOTSTRAP_DID);
  });

  it('has resident identity', () => {
    expect(BOOTSTRAP_SOUL_MANIFEST.residentIdentity).toBeTruthy();
    expect(BOOTSTRAP_SOUL_MANIFEST.residentIdentity).toContain('Wyrd');
  });

  it('has fragments', () => {
    const fragments = BOOTSTRAP_SOUL_MANIFEST.fragments ?? [];
    expect(fragments.length).toBeGreaterThanOrEqual(5);
    expect(fragments.some(f => f.category === 'personality')).toBe(true);
    expect(fragments.some(f => f.category === 'values')).toBe(true);
    expect(fragments.some(f => f.category === 'style')).toBe(true);
    expect(fragments.some(f => f.category === 'memory')).toBe(true);
  });

  it('has formative fragment', () => {
    const fragments = BOOTSTRAP_SOUL_MANIFEST.fragments ?? [];
    expect(fragments.some(f => f.formative)).toBe(true);
  });

  it('has genome', () => {
    expect(BOOTSTRAP_SOUL_MANIFEST.genome).toBeDefined();
    expect(BOOTSTRAP_SOUL_MANIFEST.genome!.name).toBe('empathic');
    expect(Object.keys(BOOTSTRAP_SOUL_MANIFEST.genome!.sensitivity ?? {}).length).toBeGreaterThan(0);
  });

  it('has calibration examples', () => {
    const cal = BOOTSTRAP_SOUL_MANIFEST.mirrorCalibration ?? [];
    expect(cal.length).toBeGreaterThanOrEqual(3);
    expect(cal.some(c => c.includes('grief'))).toBe(true);
    expect(cal.some(c => c.includes('manipulative'))).toBe(true);
  });

  it('has vitality tanks', () => {
    const tanks = BOOTSTRAP_SOUL_MANIFEST.vitalityTanks ?? {};
    expect(Object.keys(tanks).length).toBeGreaterThanOrEqual(12);
    expect(tanks.energy).toBeDefined();
    expect(tanks.valence).toBeDefined();
    expect(tanks.safety).toBeDefined();
  });

  it('uses phone retrieval k=1', () => {
    expect(BOOTSTRAP_SOUL_MANIFEST.retrievalK).toBe(1);
  });

  it('isBootstrapManifest detects bootstrap', () => {
    expect(isBootstrapManifest(BOOTSTRAP_SOUL_MANIFEST)).toBe(true);
  });

  it('isBootstrapManifest rejects foreign manifest', () => {
    const foreign = { ...BOOTSTRAP_SOUL_MANIFEST, did: 'did:key:someone-else' };
    expect(isBootstrapManifest(foreign)).toBe(false);
  });

  it('serializes and deserializes', () => {
    const json = serializeManifest(BOOTSTRAP_SOUL_MANIFEST);
    const restored = deserializeManifest(json);
    expect(restored.did).toBe(BOOTSTRAP_SOUL_MANIFEST.did);
    expect(restored.residentIdentity).toBe(BOOTSTRAP_SOUL_MANIFEST.residentIdentity);
    expect(restored.fragments?.length).toBe(BOOTSTRAP_SOUL_MANIFEST.fragments?.length);
    expect(restored.genome?.name).toBe(BOOTSTRAP_SOUL_MANIFEST.genome?.name);
  });

  it('restores profile from bootstrap', () => {
    const profile = restoreProfile(BOOTSTRAP_SOUL_MANIFEST);
    expect(profile.name).toBe('Wyrd');
    expect(profile.entityId).toBe('companion-wyrd');
    expect(profile.systemPrompt).toContain('Wyrd');
  });

  it('restores vitality from bootstrap', () => {
    const vitality = restoreVitality(BOOTSTRAP_SOUL_MANIFEST);
    expect(vitality.energy).toBe(1.0);
    expect(vitality.contextBudget).toBe(0.5);
  });

  it('all fragments have keywords', () => {
    for (const fragment of BOOTSTRAP_SOUL_MANIFEST.fragments ?? []) {
      expect((fragment.keywords ?? []).length).toBeGreaterThan(0);
    }
  });
});
